candle: write user settings to ~/.config/candle, not ${candle}/bin

gmfawcett opened the pull request:
Candle attempts to write user-specific settings into the application
directory (`${candle}/bin`), which is obviously a non-starter for Nix.
Worse, the inability to write settings causes a bad initial UI state,
where the path-visualization area is completely black (because all the
customizable colors, including the background, are incorrectly set to
black). The colors must be readjusted on every execution of the program.

This minor patch sends the settings to `~/.config/candle/settings.ini`
instead, and thereby solves both issues. (In the unlikely case that 
`HOME` is not set, `HOME=/tmp` is used as a fallback.)

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

To correct an unusable storage location for user settings, and make the program work correctly
out of the box.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

gmfawcett committed:
candle: write user settings to ~/.config/candle, not ${candle}/bin

Candle attempts to write user-specific settings into the application
directory (`${candle}/bin`), which is obviously a non-starter for Nix.
Worse, the inability to write settings causes a bad initial UI state,
where the path-visualization area is completely black (because all the
customizable colors, including the background, are incorrectly set to
black). The colors must be readjusted on every execution of the program.

This patch sends the settings to `~/.config/candle/settings.ini`
instead, and thereby solves the all-black display issue. (In the
unlikely case that `$HOME` is not set, `HOME=/tmp` is used as a
fallback.)

aanderse commented on pkgs/applications/misc/candle/candle-settings.diff:
```diff
@@ -0,0 +1,7 @@
+--- a/frmmain.cpp	1969-12-31 19:00:01.000000000 -0500
++++ b/frmmain.cpp	2021-08-16 22:17:17.502429125 -0400
+@@ -88,1 +88,3 @@
+-    m_settingsFileName = qApp->applicationDirPath() + "/settings.ini";
```
Isn't this a bug in the application? If you ran this on any other Linux distro wouldn't it try to write a directory the user doesn't have access to write to? 🤔

gmfawcett replied:
I take your point... but I've used Candle on several machines (Windows, and various Debian derivatives), and never had to _think_ about where the settings file was, the software just worked as expected. I have to assume they are patching it downstream.

An upstream fix seems unlikely -- the [git repo](https://github.com/Denvi/Candle/) has 285 open issues, and 18 open pull requests dating back a few years.  It's handy software for people driving CNC's, it's a shame that it isn't getting more attention.

aanderse replied:
Fair. It makes sense this would work on Windows, depending on how the application was installed and ran. Any chance you can look into what those other distros do here? No worries if that sounds like too much of a hassle... :smile:

I think what you wanted to do was something like this:
```cxx
// follow XDG standards
m_settingsFileName = getenv("XDG_CONFIG_HOME") ? std::string(getenv("XDG_CONFIG_HOME")) : (getenv("HOME") ? (std::string(getenv("HOME")) + "/.config") : std::string("/tmp"));
m_settingsFileName.append("/candle/settings.ini");
```

Pardon my sloppy `c++`.

You should also:
- Drop the `qDebug`, which I assume you had in there for testing.
- Include a comment which explains the situation in full.
- ping @matti-kariluoma as the maintainer to make sure they're good with this change, or have any insight on the problem.

Thanks for contributing this @gmfawcett :tada:

gmfawcett replied:
Thanks very much for the feedback.  No worries about your C++, mine is just as sloppy. Note that these are QStrings (Qt), not std::strings -- but I can work in the XDG_CONFIG_HOME variant, I forgot about that one. :)

I suggest that we leave the debug statement. The application is already debug-noisy, and this particular piece of information (where is my config being stored) could be useful to the user, esp. given that this is a downstream patch. But I don't feel strongly about it, I will remove it if that's still your preference.

I will be offline for the next few days but will follow up next week on the actions you suggested. Thanks again for your review. :)

aanderse replied:
That all sounds great :+1:

lukegb replied:
My hunch would be that something like

```c++
m_settingsFileName = QStandardPaths::locate(QStandardPaths::ConfigLocation, "settings.ini");
if (m_settingsFileName.isEmpty()) {
  m_settingsFileName = QStandardPaths::writableLocation(QStandardPaths::ConfigLocation) + "/settings.ini";
}
```

would be better... but since being properly cross-platform might not be a priority here so I'm not too fussed. It would be nice if we could have it work properly on Darwin, though, I guess.

aanderse replied:
It's been years since I've written any Qt code and I forgot about that one! Good catch.

gmfawcett replied:
@lukegb  thanks! I've updated the PR to use that approach.  Works well for me on Linux, I can't test on Darwin unfortunately.

matti-kariluoma commented:
Nothing to add to this discussion, and I relinquish my role as 
maintainer, nominating @gmfawcett

On Aug 18 2021 09:58, Aaron Andersen wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In pkgs/applications/misc/candle/candle-settings.diff 
> <https://github.com/NixOS/nixpkgs/pull/134605#discussion_r691322388>:
>
> > @@ -0,0 +1,7 @@
> +--- a/frmmain.cpp	1969-12-31 19:00:01.000000000 -0500
> ++++ b/frmmain.cpp	2021-08-16 22:17:17.502429125 -0400
> +@@ -88,1 +88,3 @@
> +-    m_settingsFileName = qApp->applicationDirPath() + "/settings.ini";
>
> That all sounds great 👍
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/NixOS/nixpkgs/pull/134605#discussion_r691322388>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AACEKPBUQYDEUFP7SWPHQ6DT5PDC7ANCNFSM5CLHU7TA>.
> Triage notifications on the go with GitHub Mobile for iOS 
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> 
> or Android 
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.
>

gmfawcett commented:
@matti-kariluoma thanks for your contributions re: Candle in Nix.  I'm happy to take responsibility for the package.

gmfawcett committed:
candle: follow XDG standards for config directory; change maintainer

@matti-kariluoma wishes to give up maintenance of candle; see
https://github.com/NixOS/nixpkgs/pull/134605#issuecomment-901462507 .
I'm willing to take responsibility for it.

gmfawcett commented:
I will be offline until next week, happy to discuss this further when I'm back.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 134605` at ba16639f run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>candle</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Consider adding a comment explaining the purpose of this patch on the line preceeding.
Near pkgs/applications/misc/candle/default.nix:14:15:
```
   |
14 |   patches = [ ./candle-settings.diff ];
   |               ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, please check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/misc/candle/default.nix:32:5:
```
   |
32 |     license = licenses.gpl3;
   |     ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/applications/misc/candle/default.nix:
```diff
@@ -24,9 +26,10 @@ mkDerivation rec {
 
   meta = with lib; {
-    description = "GRBL controller application with G-Code visualizer written in Qt";
+    description =
+      "GRBL controller application with G-Code visualizer written in Qt";
```
```suggestion
    description = "GRBL controller application with G-Code visualizer written in Qt";
```

gmfawcett replied:
@SuperSandro2000 is there an official style guide for nixpkgs patches?  I believe that my editor used `nixpkgs-fmt` to reformat that file. I assumed that it was an approved formatter?

SuperSandro2000 replied:
nixpkgs-fmt does not do such things. nixfmt is doing stupidly short lines.

gmfawcett committed:
candle: Use Qt's StandardPaths to compute config-storage location

aanderse commented on pkgs/applications/misc/candle/default.nix:
```diff
@@ -5,12 +5,14 @@ mkDerivation rec {
     sha256 = "1gpx08gdz8awbsj6lsczwgffp19z3q0r2fvm72a73qd9az29pmm0";
   };
 
+  patches = [ ./candle-settings.diff ];
```
Please include comment on why we have this patch. Mention that upstream has gone quiet, etc...

gmfawcett replied:
@aanderse Sure.  Please forgive the newbie questions that follow -- I don't spend much time in Github PR Land. :) 
I don't think I can rebase a PR (can I?), so did you want me to git-revert my changes, and then make the new commits you indicated?  Or something else -- e.g., should each of the three changes be in its own separate PR?

aanderse reviewed (changes requested):
Please split this into 3 commits as follows:
1. adding gmfawcett to maintainer-list.nix
2. change maintainer from matti-kariluoma to gmfawcett
3. add patch

gmfawcett committed:
Merge branch 'NixOS:master' into candle-settings-relocate
